File: mcp/src/tools/desktop/fxagent-client.ts
export const DEFAULT_HOST = '127.0.0.1';
export const DEFAULT_PORT = 4567;
const REQUEST_TIMEOUT_MS = 10_000;

export class FxAgentConnectionError extends Error {
  constructor(host: string, port: number) {
    super(
      `Cannot connect to fxagent at ${host}:${port} — ` +
      `is the app running with -javaagent:fxagent.jar?`,
    );
    this.name = 'FxAgentConnectionError';
  }
}

export class FxAgentProtocolError extends Error {
  constructor(
    public readonly status: number,
    endpoint: string,
    detail?: string,
  ) {
    super(
      `fxagent returned HTTP ${status} for ${endpoint}` +
      (detail ? `: ${detail}` : ''),
    );
    this.name = 'FxAgentProtocolError';
  }
}

// ---- Raw response shapes from fxagent ----

export interface SceneTreeResponse {
  windows: Array<{
    windowIndex: number;
    windowType: string;
    root: ElementNode | Record<string, never>;
  }>;
}

export interface ElementNode {
  handle: string;
  id: string | null;
  type: string;
  fullType: string;
  styleClasses: string[];
  text: string | null;
  bounds: { x: number; y: number; width: number; height: number } | null;
  visible: boolean;
  enabled: boolean;
  focused: boolean;
  properties: Record<string, unknown> | null;
  children?: ElementNode[];
}

export interface QueryResponse {
  elements: ElementNode[];
  count: number;
}

export interface ActionResponse {
  success: boolean;
  message: string;
  element?: ElementNode;
}

export interface ScreenshotResponse {
  data: string;   // base64-encoded PNG (field name confirmed from ScreenshotController.java)
  format: string;
  width: number;
  height: number;
}

// ---- Client ----

export class FxAgentClient {
  private readonly baseUrl: string;
  private readonly host: string;
  private readonly port: number;

  constructor(host: string = DEFAULT_HOST, port: number = DEFAULT_PORT) {
    this.host = host;
    this.port = port;
    this.baseUrl = `http://${host}:${port}`;
  }

  async getSceneTree(depth: number): Promise<SceneTreeResponse> {
    return this.get<SceneTreeResponse>(`/api/v1/scene/tree?depth=${depth}`);
  }

  async getElementSubtree(handle: string, depth: number): Promise<ElementNode> {
    return this.get<ElementNode>(
      `/api/v1/elements/${encodeURIComponent(handle)}/tree?depth=${depth}`,
    );
  }

  async queryElements(
    selector: string,
    maxResults: number,
    windowIndex?: number,
  ): Promise<QueryResponse> {
    const body: Record<string, unknown> = { selector, maxResults };
    if (windowIndex !== undefined) body.windowIndex = windowIndex;
    return this.post<QueryResponse>('/api/v1/elements/query', body);
  }

  async performAction(
    selector: string,
    action: string,
    value?: string,
  ): Promise<ActionResponse> {
    const body: Record<string, unknown> = { selector, action };
    if (value !== undefined) body.value = value;
    return this.post<ActionResponse>('/api/v1/actions', body);
  }

  async waitForElement(
    selector: string,
    condition: string,
    timeoutMs: number,
  ): Promise<ElementNode> {
    return this.post<ElementNode>('/api/v1/elements/wait', {
      selector,
      condition,
      timeoutMs,
      pollIntervalMs: 200,
    });
  }

  /**
   * Poll `/api/v1/elements/query` client-side until `selector` reaches `state`.
   *
   * fxagent's own `/api/v1/elements/wait` can only ever *find* an element: when
   * nothing matches it answers HTTP 500 (`INTERNAL_ERROR … parameter obj is
   * null`), so it cannot express `hidden`/`disabled`/absent. Verified against
   * fxagent 0.3.0. The npm `javafx-driver` package polls the query endpoint for
   * exactly this reason — we mirror that here.
   *
   * @returns the matching element, or `null` for states satisfied by absence.
   */
  async waitForState(
    selector: string,
    state: 'visible' | 'hidden' | 'enabled' | 'disabled' | 'exists',
    timeoutMs: number,
    pollIntervalMs = 200,
  ): Promise<{ satisfied: boolean; element: ElementNode | null }> {
    const deadline = Date.now() + timeoutMs;
    let last: ElementNode | null = null;

    for (;;) {
      last = await this.firstMatch(selector);
      if (matchesState(last, state)) {
        return { satisfied: true, element: last };
      }
      if (Date.now() >= deadline) {
        return { satisfied: false, element: last };
      }
      await delay(Math.min(pollIntervalMs, Math.max(0, deadline - Date.now())));
    }
  }

  /** Single-element query; returns null when the selector matches nothing. */
  private async firstMatch(selector: string): Promise<ElementNode | null> {
    try {
      const response = await this.queryElements(selector, 1);
      return response.elements?.[0] ?? null;
    } catch (err) {
      // A selector that matches nothing may surface as a protocol error
      // depending on the agent build — treat it as "not present".
      if (err instanceof FxAgentProtocolError) return null;
      throw err;
    }
  }

  async captureScreenshot(opts: {
    selector?: string;
    windowIndex?: number;
  }): Promise<ScreenshotResponse> {
    return this.post<ScreenshotResponse>('/api/v1/screenshot', opts);
  }

  private async get<T>(path: string): Promise<T> {
    return this.request<T>('GET', path);
  }

  private async post<T>(path: string, body: unknown): Promise<T> {
    return this.request<T>('POST', path, body);
  }

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);

    try {
      const headers: Record<string, string> = {};
      let requestBody: string | undefined;
      if (body !== undefined) {
        headers['Content-Type'] = 'application/json';
        requestBody = JSON.stringify(body);
      }

      const response = await fetch(url, {
        method,
        headers,
        body: requestBody,
        signal: controller.signal,
      });

      const responseText = await response.text();
      let parsed: unknown;
      try {
        parsed = JSON.parse(responseText);
      } catch {
        parsed = responseText;
      }

      if (!response.ok) {
        const detail =
          typeof parsed === 'object' && parsed !== null && 'error' in parsed
            ? String((parsed as { error: unknown }).error)
            : typeof parsed === 'string' && parsed.length > 0
              ? parsed.slice(0, 200)
              : undefined;
        throw new FxAgentProtocolError(response.status, path, detail);
      }

      return parsed as T;
    } catch (err) {
      if (err instanceof FxAgentProtocolError) throw err;

      const e = err as NodeJS.ErrnoException & { name?: string };
      if (
        e.name === 'AbortError' ||
        e.code === 'ECONNREFUSED' ||
        e.code === 'ECONNRESET' ||
        e.code === 'UND_ERR_CONNECT_TIMEOUT' ||
        (e.cause instanceof Error && (e.cause as NodeJS.ErrnoException).code === 'ECONNREFUSED')
      ) {
        throw new FxAgentConnectionError(this.host, this.port);
      }
      throw err;
    } finally {
      clearTimeout(timer);
    }
  }
}

/** Whether a query result satisfies a wait state. */
function matchesState(
  element: ElementNode | null,
  state: 'visible' | 'hidden' | 'enabled' | 'disabled' | 'exists',
): boolean {
  switch (state) {
    case 'visible':
      return element !== null && element.visible;
    case 'hidden':
      // Satisfied both by an absent node and by a present-but-invisible one.
      return element === null || !element.visible;
    case 'enabled':
      return element !== null && element.visible && element.enabled;
    case 'disabled':
      return element !== null && !element.enabled;
    case 'exists':
      return element !== null;
  }
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: java/conductor-core/src/main/java/com/nouhouari/conductor/drivers/internal/FxAgentClient.java
package com.nouhouari.conductor.drivers.internal;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Plain HTTP/JSON client for the {@code fxagent.jar} Java agent that
 * instruments JavaFX apps under test (started via {@code -javaagent:fxagent.jar}).
 *
 * <p>The wire protocol was reverse-engineered from this repo's own
 * {@code mcp/src/tools/desktop/fxagent-client.ts} (an MCP tool that talks to
 * the same agent) rather than fxagent's source, which isn't vendored in this
 * repo. It's a plain REST/JSON API on {@code 127.0.0.1:4567} by default, so
 * porting it needed no reverse-engineering of the npm {@code javafx-driver}
 * package that the TypeScript side wraps.
 */
public class FxAgentClient {

    public static final String DEFAULT_HOST = "127.0.0.1";
    public static final int DEFAULT_PORT = 4567;
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(10);

    public record ElementNode(String handle, String id, String type, String fullType, boolean visible,
            boolean enabled) {
    }

    public record QueryResponse(java.util.List<ElementNode> elements) {
    }

    public record ActionResponse(boolean success, String message) {
    }

    public record ScreenshotResponse(String data, String format, int width, int height) {
    }

    public static class ConnectionException extends RuntimeException {
        public ConnectionException(String host, int port, Throwable cause) {
            super("Cannot connect to fxagent at " + host + ":" + port
                    + " — is the app running with -javaagent:fxagent.jar?", cause);
        }
    }

    public static class ProtocolException extends RuntimeException {
        public ProtocolException(int status, String endpoint, String detail) {
            super("fxagent returned HTTP " + status + " for " + endpoint + (detail != null ? ": " + detail : ""));
        }
    }

    private final String baseUrl;
    private final String host;
    private final int port;
    private final HttpClient http = HttpClient.newHttpClient();
    private final ObjectMapper json = new ObjectMapper()
            .disable(com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);

    public FxAgentClient() {
        this(DEFAULT_HOST, DEFAULT_PORT);
    }

    public FxAgentClient(String host, int port) {
        this.host = host;
        this.port = port;
        this.baseUrl = "http://" + host + ":" + port;
    }

    public boolean isReady() {
        try {
            request("GET", "/api/v1/scene/tree?depth=0", null);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public ActionResponse performAction(String selector, String action, String value) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("selector", selector);
        body.put("action", action);
        if (value != null) {
            body.put("value", value);
        }
        return post("/api/v1/actions", body, ActionResponse.class);
    }

    public ElementNode waitForElement(String selector, String condition, int timeoutMs) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("selector", selector);
        body.put("condition", condition);
        body.put("timeoutMs", timeoutMs);
        body.put("pollIntervalMs", 200);
        return post("/api/v1/elements/wait", body, ElementNode.class);
    }

    /**
     * Single-element query used for client-side wait polling — the agent's own
     * {@code /elements/wait} cannot express "hidden"/"absent" (it fails when no
     * element matches), so waits are polled here exactly as the npm
     * {@code javafx-driver} does.
     */
    public QueryResponse queryNode(String selector) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("selector", selector);
        body.put("maxResults", 1);
        return post("/api/v1/elements/query", body, QueryResponse.class);
    }

    public ScreenshotResponse captureScreenshot(String selector, Integer windowIndex) {
        Map<String, Object> body = new LinkedHashMap<>();
        if (selector != null) {
            body.put("selector", selector);
        }
        if (windowIndex != null) {
            body.put("windowIndex", windowIndex);
        }
        return post("/api/v1/screenshot", body, ScreenshotResponse.class);
    }

    public static byte[] decodePng(ScreenshotResponse response) {
        return Base64.getDecoder().decode(response.data());
    }

    private <T> T post(String path, Object body, Class<T> type) {
        return request("POST", path, body, type);
    }

    private <T> T request(String method, String path, Object body, Class<T> type) {
        String raw = request(method, path, body);
        try {
            return json.readValue(raw, type);
        } catch (IOException e) {
            throw new RuntimeException("Failed to parse fxagent response for " + path, e);
        }
    }

    private String request(String method, String path, Object body) {
        try {
            HttpRequest.Builder builder = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .timeout(REQUEST_TIMEOUT);

            if (body != null) {
                String payload = json.writeValueAsString(body);
                builder.header("Content-Type", "application/json")
                        .method(method, HttpRequest.BodyPublishers.ofString(payload));
            } else {
                builder.method(method, HttpRequest.BodyPublishers.noBody());
            }

            HttpResponse<String> response = http.send(builder.build(), HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() >= 400) {
                throw new ProtocolException(response.statusCode(), path, response.body());
            }
            return response.body();
        } catch (java.net.ConnectException e) {
            throw new ConnectionException(host, port, e);
        } catch (IOException e) {
            throw new RuntimeException("fxagent request failed: " + method + " " + path, e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException("fxagent request interrupted: " + method + " " + path, e);
        }
    }
}
